src/Microsoft.ML.Core/Data/IValueMapper.cs

```diff
@@ -9,13 +9,13 @@ namespace Microsoft.ML.Runtime.Data
     /// <summary>
     /// Delegate type to map/convert a value.
     /// </summary>
-    public delegate void ValueMapper<TSrc, TDst>(ref TSrc src, ref TDst dst);
+    public delegate void ValueMapper<TSrc, TDst>(in TSrc src, ref TDst dst);
 
     /// <summary>
     /// Delegate type to map/convert among three values, for example, one input with two
     /// outputs, or two inputs with one output.
     /// </summary>
-    public delegate void ValueMapper<TVal1, TVal2, TVal3>(ref TVal1 val1, ref TVal2 val2, ref TVal3 val3);
+    public delegate void ValueMapper<TVal1, TVal2, TVal3>(in TVal1 val1, ref TVal2 val2, ref TVal3 val3);
```

Comment:
I inspected all the usages of this delegate, and most times it is used by the below interface `IValueMapperDist`, which takes 1 input and returns 2 outputs.
The only case where this was used for 2 inputs and 1 output was in one place in matrix factorization code (the two inputs were column and row I believe).

I didn't feel it was necessary to split this delegate. If someone does feel it is necessary, I think a separate issue should be logged and addressed separately.